optuna/storages/_rdb/storage.py

```diff
@@ -446,46 +468,50 @@ def _create_new_trial(
+            with _create_scoped_session(self.scoped_session, ignore_timeout_error=True) as session:
                 try:
                     # Ensure that that study exists.
                     #
                     # Locking within a study is necessary since the creation of a trial is not an
                     # atomic operation. More precisely, the trial number computed in
                     # `_get_prepared_new_trial` is prone to race conditions without this lock.
                     models.StudyModel.find_or_raise_by_id(study_id, session, for_update=True)
-
                     trial = self._get_prepared_new_trial(study_id, template_trial, session)
+                    if template_trial:
+                        frozen = copy.deepcopy(template_trial)
+                        frozen.number = trial.number
+                        frozen.datetime_start = trial.datetime_start
+                        frozen._trial_id = trial.trial_id
+                    else:
+                        frozen = FrozenTrial(
+                            number=trial.number,
+                            state=trial.state,
+                            value=None,
+                            values=None,
+                            datetime_start=trial.datetime_start,
+                            datetime_complete=None,
+                            params={},
+                            distributions={},
+                            user_attrs={},
+                            system_attrs={},
+                            intermediate_values={},
+                            trial_id=trial.trial_id,
+                        )
                     break  # Successfully created trial.
                 except sqlalchemy_exc.OperationalError:
-                    if n_retries > 2:
-                        raise
+                    # If this is a timeout error, `_create_scoped_session` will catch it
+                    # and the outer `while` will retry this transaction after backoff.
+                    raise
+            n_retries += 1
 
-                n_retries += 1
-
-            if template_trial:
-                frozen = copy.deepcopy(template_trial)
-                frozen.number = trial.number
-                frozen.datetime_start = trial.datetime_start
-                frozen._trial_id = trial.trial_id
-            else:
-                frozen = FrozenTrial(
-                    number=trial.number,
-                    state=trial.state,
-                    value=None,
-                    values=None,
-                    datetime_start=trial.datetime_start,
-                    datetime_complete=None,
-                    params={},
-                    distributions={},
-                    user_attrs={},
-                    system_attrs={},
-                    intermediate_values={},
-                    trial_id=trial.trial_id,
-                )
+        assert n_retries <= MAX_RETRIES, "The _create_new_trial was repeatedly failed."
```

Comment:
I think it is more appropriate to use `raise RuntimeError` or something similar, but not `assert`.

Reply:
Agreed. I will revise. Fow now, I think that `raise sqlalchemy_exc.OperationalError` is appropriate.